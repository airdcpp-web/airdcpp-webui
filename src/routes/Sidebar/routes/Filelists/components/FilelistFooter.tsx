import * as React from 'react';

import { formatSize } from 'utils/ValueFormat';
import { FooterItem, SessionFooter } from 'routes/Sidebar/components/SessionFooter';
import { useMobileLayout } from 'utils/BrowserUtils';

import * as API from 'types/api';
import * as UI from 'types/ui';

interface FilelistFooterProps {
  session: API.FilelistSession;
  sessionT: UI.ModuleTranslator;
}

const FilelistFooter: React.FC<FilelistFooterProps> = ({ session, sessionT }) => {
  if (useMobileLayout()) {
    return null;
  }

  let locationText = session.location!.type.str;
  if (locationText.length > 0) {
    locationText = `${formatSize(
      session.location!.size,
      sessionT.plainT,
    )} (${locationText})`;
  }

  return (
    <SessionFooter>
      <FooterItem label={sessionT.translate('Directory size')} text={locationText} />
      <FooterItem
        label={sessionT.translate('Total list size')}
        text={formatSize(session.total_size, sessionT.plainT)}
      />
    </SessionFooter>
  );
};

export default FilelistFooter;
