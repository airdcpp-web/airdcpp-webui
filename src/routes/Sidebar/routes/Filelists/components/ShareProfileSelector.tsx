import Dropdown from '@/components/semantic/Dropdown';
import MenuItemLink from '@/components/semantic/MenuItemLink';

import ShareProfileDecorator, {
  ShareProfileDecoratorChildProps,
} from '@/decorators/ShareProfileDecorator';

import * as API from '@/types/api';
import * as UI from '@/types/ui';

import { formatProfileNameWithSize } from '@/utils/ShareProfileUtils';
import { useFormatter } from '@/context/FormatterContext';

interface ShareProfileSelectorProps {
  // Callback after selecting a profile
  onProfileChanged: (profileId: number) => void;

  sessionT: UI.ModuleTranslator;
}

const ShareProfileSelector: React.FC<
  ShareProfileSelectorProps & ShareProfileDecoratorChildProps
> = ({ sessionT, profiles, onProfileChanged }) => {
  const formatter = useFormatter();

  const getDropdownItem = (profile: API.ShareProfile) => {
    return (
      <MenuItemLink key={profile.id} onClick={() => onProfileChanged(profile.id)}>
        {formatProfileNameWithSize(profile, formatter)}
      </MenuItemLink>
    );
  };

  return (
    <Dropdown
      className="profile top right pointing"
      caption={sessionT.translate('Browse own share...')}
      triggerIcon=""
    >
      {profiles.map(getDropdownItem)}
    </Dropdown>
  );
};

export default ShareProfileDecorator(ShareProfileSelector, false);
