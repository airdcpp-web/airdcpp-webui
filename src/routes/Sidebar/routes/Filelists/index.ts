import IconConstants from 'constants/IconConstants';
import FilelistSessionStore from 'stores/FilelistSessionStore';

import * as API from 'types/api';
import * as UI from 'types/ui';

export const FilelistRoutes: UI.RouteItem = {
  title: 'Filelists',
  path: '/filelists',
  matchPath: '/filelists/:session?/:id?',
  icon: IconConstants.FILELISTS_COLORED,
  unreadInfoStore: FilelistSessionStore,
  access: API.AccessEnum.FILELISTS_VIEW,
  // lazy: () => import(/* webpackChunkName: "filelists" */ './components/Filelists'),
  lazy: async () => {
    const { default: Component } = await import(
      /* webpackChunkName: "filelists" */ './components'
    );
    return {
      Component,
    };
  },
};
