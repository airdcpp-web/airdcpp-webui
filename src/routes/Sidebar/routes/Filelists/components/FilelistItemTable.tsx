import { useMemo } from 'react';
import * as React from 'react';

import FilelistItemActions from 'actions/ui/filelist/FilelistItemActions';

import { dupeToStringType } from 'utils/TypeConvert';
import { TableActionMenu } from 'components/action-menu';

import FilelistViewStore from 'stores/FilelistViewStore';
import FilelistSessionStore from 'stores/FilelistSessionStore';

import VirtualTable from 'components/table/VirtualTable';
import {
  SizeCell,
  DurationCell,
  FileDownloadCell,
  FileDownloadCellClickHandler,
} from 'components/table/Cell';
import { Column } from 'fixed-data-table-2';

import IconConstants from 'constants/IconConstants';
import Loader from 'components/semantic/Loader';
import Message from 'components/semantic/Message';

import * as API from 'types/api';
import * as UI from 'types/ui';

import { FilelistItemInfoDialog } from './item-info-dialog';
import { RowWrapperCellChildProps } from 'components/table/RowWrapperCell';
import { filelistDownloadHandler } from 'services/api/FilelistApi';
import MenuConstants from 'constants/MenuConstants';

interface NameCellProps extends RowWrapperCellChildProps<string, API.FilelistItem> {
  session: API.FilelistSession;
  onClickDirectory: FileDownloadCellClickHandler;
}

const NameCell: React.FC<NameCellProps> = ({
  rowDataGetter,
  onClickDirectory,
  session,
  ...other
}) => {
  const itemDataGetter = useMemo(() => {
    return () => ({
      item: rowDataGetter!(),
      session,
    });
  }, [session]);

  return (
    <FileDownloadCell
      clickHandlerGetter={onClickDirectory}
      userGetter={() => session.user}
      downloadHandler={filelistDownloadHandler}
      rowDataGetter={rowDataGetter}
      session={session}
      entityId={session.id}
      remoteMenuId={MenuConstants.FILELIST_ITEM}
      {...other}
    >
      <TableActionMenu
        actions={FilelistItemActions}
        itemData={itemDataGetter}
        ids={['refreshShare', 'details']}
      />
    </FileDownloadCell>
  );
};

interface ListBrowserProps {
  session: API.FilelistSession;
  onClickDirectory: (path: string) => void;
  sessionT: UI.ModuleTranslator;
}

class FilelistItemTable extends React.Component<ListBrowserProps> {
  rowClassNameGetter = (rowData: API.FilelistItem) => {
    // Don't highlight dupes in own filelist...
    const isOwnList = this.props.session.user.flags.includes('self');
    return isOwnList ? '' : dupeToStringType(rowData.dupe);
  };

  emptyRowsNodeGetter = () => {
    const { location, state } = this.props.session;
    const { translate } = this.props.sessionT;

    if (state.id === 'download_failed') {
      return (
        <Message
          icon={IconConstants.ERROR}
          title={translate('Download failed')}
          description={state.str}
        />
      );
    }

    // The list finished downloading but the view hasn't updated yet
    const { files, directories } = location!.type as API.DirectoryType;
    if (files !== 0 || directories !== 0) {
      return <Loader text={translate('Updating view')} />;
      //return null;
    }

    // The directory was changed but the download state hasn't changed yet
    if (!location!.complete) {
      return <Loader text={translate('Preparing download')} />;
    }

    return (
      <Message
        title={translate('No content to display')}
        description={translate('The directory is empty')}
      />
    );
  };

  onClickDirectory: FileDownloadCellClickHandler = (cellData, rowDataGetter) => {
    if (rowDataGetter().type.id === 'directory') {
      return () =>
        this.props.onClickDirectory(this.props.session.location!.path + cellData + '/');
    }

    return undefined;
  };

  userGetter = () => {
    return this.props.session.user;
  };

  render() {
    const { session } = this.props;
    return (
      <>
        <VirtualTable
          emptyRowsNodeGetter={this.emptyRowsNodeGetter}
          rowClassNameGetter={this.rowClassNameGetter}
          store={FilelistViewStore}
          entityId={session.id}
          viewId={session.location!.path}
          sessionStore={FilelistSessionStore}
          moduleId={FilelistItemActions.moduleId}
        >
          <Column
            name="Name"
            width={200}
            columnKey="name"
            cell={<NameCell onClickDirectory={this.onClickDirectory} session={session} />}
            flexGrow={8}
          />
          <Column
            name="Size"
            width={60}
            columnKey="size"
            cell={<SizeCell />}
            flexGrow={1}
          />
          <Column name="Type" width={70} columnKey="type" flexGrow={1} hideWidth={500} />
          <Column
            name="Last modified"
            width={80}
            columnKey="time"
            cell={<DurationCell />}
            flexGrow={1}
          />
        </VirtualTable>
        <FilelistItemInfoDialog session={session} />
      </>
    );
  }
}

export default FilelistItemTable;
