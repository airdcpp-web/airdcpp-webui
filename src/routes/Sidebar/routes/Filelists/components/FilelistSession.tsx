import { Component } from 'react';

import ListBrowser, { FilelistLocationState } from 'routes/Sidebar/routes/Filelists/components/ListBrowser';
import FilelistFooter from 'routes/Sidebar/routes/Filelists/components/FilelistFooter';

import ActiveSessionDecorator from 'decorators/ActiveSessionDecorator';

import Loader from 'components/semantic/Loader';
import Message from 'components/semantic/Message';

import * as API from 'types/api';
import * as UI from 'types/ui';

import { SessionChildProps } from 'routes/Sidebar/components/SessionLayout';
import { Location } from 'history';


type FilelistSessionProps = SessionChildProps<API.FilelistSession, UI.EmptyObject, UI.EmptyObject>

class FilelistSession extends Component<FilelistSessionProps> {
  static displayName = 'FilelistSession';

  render() {
    const { session, sessionT } = this.props;
    const { user, location, state } = session;

    if (user.flags.indexOf('offline') !== -1 && user.flags.indexOf('self') === -1) {
      return (
        <div className="filelist session">
          <Message 
            title={ sessionT.t('userOffline', 'User offline') }
            description={ sessionT.t<string>(
              'userOfflineDesc', 
              'You will be able to continue browsing when the user comes back online'
            ) }
          />
        </div>
      );
    }

    if ((state.id !== 'loaded' && state.id !== 'download_failed') || !location) {
      return (
        <div className="filelist session">
          <Loader text={ state.str }/>
        </div>
      );
    }

    return (
      <div className="filelist session">
        <ListBrowser
          location={ this.props.location as Location<FilelistLocationState> }
          session={ session }
          sessionT={ sessionT }
        />

        <FilelistFooter
          session={ session }
          sessionT={ sessionT }
        />
      </div>
    );
  }
}

export default ActiveSessionDecorator<FilelistSessionProps, API.FilelistSession>(FilelistSession);
