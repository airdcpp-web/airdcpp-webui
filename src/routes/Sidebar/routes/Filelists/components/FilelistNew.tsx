import RecentLayout from 'routes/Sidebar/components/RecentLayout';

import FilelistSessionActions from 'actions/reflux/FilelistSessionActions';
import FilelistSessionStore from 'stores/FilelistSessionStore';

import ShareProfileSelector from 'routes/Sidebar/routes/Filelists/components/ShareProfileSelector';
import { HistoryEntryEnum } from 'constants/HistoryConstants';

import * as API from 'types/api';

import IconConstants from 'constants/IconConstants';
import { UserSelectField } from 'components/select';
import { NewSessionLayoutProps } from 'routes/Sidebar/components/types';

const FilelistNew: React.FC<NewSessionLayoutProps> = (props) => {
  const handleSubmit = (user: API.HintedUser) => {
    const { navigate, location } = props;
    FilelistSessionActions.createSession(
      { user },
      {
        navigate,
        location,
        sessionStore: FilelistSessionStore,
      }
    );
  };

  const onProfileChanged = (profileId: number) => {
    const { navigate, location } = props;
    FilelistSessionActions.ownList(profileId, {
      navigate,
      location,
      sessionStore: FilelistSessionStore,
    });
  };

  const recentUserRender = (entry: API.HistoryItem) => {
    return (
      <a onClick={(_) => handleSubmit(entry.user!)}>
        {entry.user!.nicks + (entry.description ? ` (${entry.description})` : '')}
      </a>
    );
  };

  const hasSession = (entry: API.HistoryItem) => {
    return FilelistSessionStore.getSession(entry.user!.cid);
  };

  const { sessionT } = props;
  return (
    <div className="filelist session new">
      <UserSelectField
        onChange={handleSubmit}
        offlineMessage={sessionT.t<string>(
          'offlineMessage',
          'You must to be connected to at least one hub in order to download filelists from other users'
        )}
        isClearable={false}
        autoFocus={true}
      />
      <ShareProfileSelector onProfileChanged={onProfileChanged} sessionT={sessionT} />
      <RecentLayout
        entryType={HistoryEntryEnum.FILELIST}
        hasSession={hasSession}
        entryTitleRenderer={recentUserRender}
        entryIcon={IconConstants.FILELISTS_PLAIN}
      />
    </div>
  );
};

export default FilelistNew;
