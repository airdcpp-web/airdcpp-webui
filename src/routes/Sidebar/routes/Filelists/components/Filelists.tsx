import * as React from 'react';

import SessionLayout from '@/routes/Sidebar/components/SessionLayout';
import UserItemHandlerDecorator from '@/routes/Sidebar/decorators/UserItemHandlerDecorator';

import FilelistNew from '@/routes/Sidebar/routes/Filelists/components/FilelistNew';
import FilelistSession from '@/routes/Sidebar/routes/Filelists/components/FilelistSession';

import '../style.css';

import * as API from '@/types/api';
import * as UI from '@/types/ui';
import {
  SessionProviderDecorator,
  SessionProviderDecoratorChildProps,
} from '@/routes/Sidebar/decorators/SessionProviderDecorator';
import IconConstants from '@/constants/IconConstants';
import { FilelistSessionActionMenu } from '@/actions/ui/filelist';
import { FilelistAPIActions } from '@/actions/store/FilelistActions';
import { FilelistStoreSelector } from '@/stores/session/filelistSlice';
import MenuConstants from '@/constants/MenuConstants';

const UserItemHandler = UserItemHandlerDecorator(['message']);
const ItemHandler: UI.SessionInfoGetter<API.FilelistSession> = {
  itemLabelGetter() {
    return null;
  },

  itemNameGetter(session) {
    return session.share_profile
      ? session.share_profile.str
      : UserItemHandler.itemNameGetter(session);
  },

  itemStatusGetter(session) {
    return session.share_profile ? 'blue' : UserItemHandler.itemStatusGetter!(session);
  },

  itemHeaderTitleGetter(session, location, actionMenu) {
    return session.share_profile
      ? actionMenu
      : UserItemHandler.itemHeaderTitleGetter!(session, location, actionMenu);
  },

  itemHeaderDescriptionGetter(session) {
    return session.share_profile
      ? null
      : UserItemHandler.itemHeaderDescriptionGetter(session);
  },

  itemHeaderIconGetter(session) {
    return session.share_profile
      ? 'green server'
      : UserItemHandler.itemHeaderIconGetter(session);
  },
};

const Filelists: React.FC<SessionProviderDecoratorChildProps<API.FilelistSession>> = (
  props,
) => {
  const { params, sessionT, ...other } = props;
  return (
    <SessionLayout
      activeId={params.id}
      baseUrl="filelists"
      newCaption={sessionT.t('new', 'Open filelist')}
      newDescription={sessionT.t('newDesc', 'Start browsing a new filelist')}
      newIcon={IconConstants.FILELISTS_PLAIN}
      disableSideMenu={true}
      editAccess={API.AccessEnum.FILELISTS_EDIT}
      uiActions={FilelistSessionActionMenu}
      sessionApi={FilelistAPIActions}
      sessionStoreSelector={FilelistStoreSelector}
      sessionItemLayout={FilelistSession}
      newLayout={FilelistNew}
      sessionT={sessionT}
      remoteMenuId={MenuConstants.FILELIST}
      {...ItemHandler}
      {...other}
    />
  );
};

export default SessionProviderDecorator(
  Filelists,
  FilelistStoreSelector,
  UI.Modules.FILELISTS,
);
