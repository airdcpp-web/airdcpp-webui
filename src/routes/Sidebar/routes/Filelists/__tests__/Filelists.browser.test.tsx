import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import { waitFor, within } from '@testing-library/react';

import { renderDataRoutes } from '@/tests/render/test-renderers';

import { CommonDataMocks, initCommonDataMocks } from '@/tests/mocks/mock-data-common';

import * as API from '@/types/api';

import SocketNotificationListener from '@/components/main/notifications/SocketNotificationListener';

import { setupUserEvent } from '@/tests/helpers/test-form-helpers';
import { VIEW_FIXED_HEIGHT } from '@/tests/render/test-containers';
import { getMockServer, MockServer } from '@/tests/mocks/mock-server';
import { createDataFetchRoutes } from '@/tests/helpers/test-route-helpers';
import FilelistConstants from '@/constants/FilelistConstants';
import {
  FilelistDirectoryDownloadResponse,
  FilelistDirectoryItemsList,
  FilelistGetFilelistItemDirectoryResponse,
  FilelistGetFilelistItemFileResponse,
  FilelistLoadedResponse,
  FilelistMeResponse,
  FilelistPendingResponse,
  FilelistRootItemsList,
  FilelistStateLoaded,
  FilelistStatePending,
} from '@/tests/mocks/api/filelist';
import Filelists from '../components/Filelists';
import { installTableMocks } from '@/tests/mocks/mock-table';
import { selectTopLayoutSession } from '@/tests/helpers/test-session-helpers';
import {
  installBasicSessionHandlers,
  installUserSearchFieldMocks,
  waitSessionsLoaded,
} from '@/tests/mocks/mock-session';
import { clickMenuItem, openIconMenu, openMenu } from '@/tests/helpers/test-menu-helpers';
import { installActionMenuMocks } from '@/components/action-menu/__tests__/test-action-menu-helpers';
import MenuConstants from '@/constants/MenuConstants';
import { RemoteMenuGrouped1 } from '@/tests/mocks/api/menu';
import QueueConstants from '@/constants/QueueConstants';
import { QueueBundleCreateFileResponse } from '@/tests/mocks/api/queue-bundles';
import {
  navigateToUrl,
  waitExpectRequestToMatchSnapshot,
  waitForUrl,
} from '@/tests/helpers/test-helpers';
import { installShareProfileMocks } from '@/tests/mocks/mock-share';
import { HistoryEntryEnum } from '@/constants/HistoryConstants';
import { HistoryFilelistResponse } from '@/tests/mocks/api/history';
import { getHistoryUrl } from '@/routes/Sidebar/components/RecentLayout';
import { ShareProfile2 } from '@/tests/mocks/api/share-profiles';
import { formatProfileNameWithSize } from '@/utils/ShareProfileUtils';

// tslint:disable:no-empty
describe('Filelists', () => {
  let server: MockServer;

  beforeEach(() => {
    server = getMockServer();
  });

  afterEach(() => {
    server.stop();
  });

  const getSocket = async () => {
    const commonData = await initCommonDataMocks(server, {
      permissions: [
        API.AccessEnum.FILELISTS_VIEW,
        API.AccessEnum.FILELISTS_EDIT,
        API.AccessEnum.DOWNLOAD,
      ],
    });

    // Messages
    const session1Mocks = installBasicSessionHandlers(
      FilelistConstants.SESSIONS_URL,
      FilelistLoadedResponse.id,
      server,
    );
    const session2Mocks = installBasicSessionHandlers(
      FilelistConstants.SESSIONS_URL,
      FilelistPendingResponse.id,
      server,
    );

    // Tables
    const filelist1TableMocks = installTableMocks(FilelistRootItemsList.items, {
      server,
      moduleUrl: FilelistConstants.MODULE_URL,
      viewName: FilelistConstants.VIEW_ID,
      entityId: FilelistLoadedResponse.id,
    });

    const filelistMeTableMocks = installTableMocks(FilelistRootItemsList.items, {
      server,
      moduleUrl: FilelistConstants.MODULE_URL,
      viewName: FilelistConstants.VIEW_ID,
      entityId: FilelistMeResponse.id,
    });

    // Menus
    const itemMenuMocks = installActionMenuMocks(
      MenuConstants.FILELIST_ITEM,
      [RemoteMenuGrouped1],
      server,
    );

    return {
      commonData,

      filelist1Mocks: {
        ...session1Mocks,
        table: filelist1TableMocks,
      },

      filelist2Mocks: {
        ...session2Mocks,
      },

      filelistMeMocks: {
        table: filelistMeTableMocks,
      },

      itemMenuMocks,
    };
  };

  const renderLayout = async () => {
    const { commonData, ...other } = await getSocket();
    const { sessionStore } = commonData;
    sessionStore.getState().activity.setUserActive(true);

    const FilelistLayoutTest = () => {
      return (
        <>
          <Filelists />
          <SocketNotificationListener />
        </>
      );
    };

    const routes = createDataFetchRoutes([
      {
        path: '/filelists/:session?/:id?/*',
        Component: FilelistLayoutTest,
      },
    ]);

    const renderData = renderDataRoutes(routes, commonData, {
      viewType: VIEW_FIXED_HEIGHT,
      routerProps: { initialEntries: ['/filelists'] },
    });

    const userEvent = setupUserEvent();
    return { ...commonData, ...renderData, ...other, userEvent };
  };

  test('should render', async () => {
    const renderResult = await renderLayout();
    const { findByText, filelist1Mocks, filelist2Mocks, router } = renderResult;

    // Check content
    await findByText(FilelistRootItemsList.items[0].name);
    // expect(filelist1Mocks.onSessionRead).toHaveBeenCalled();

    // Open a different session
    await selectTopLayoutSession(
      FilelistPendingResponse.user.nicks,
      renderResult,
      filelist2Mocks.onSessionRead,
    );

    // Check content
    await findByText('Download pending');

    expect(filelist1Mocks.onSessionRead).not.toHaveBeenCalled();

    await navigateToUrl('/', router);
    await filelist1Mocks.table.waitStopped();
  });

  test('should navigate between directories', async () => {
    const filelist = FilelistLoadedResponse;
    const clickedDirectory = FilelistGetFilelistItemDirectoryResponse;

    // Filelist handlers
    const onChangeDirectoryChild = vi.fn();
    server.addRequestHandler(
      'POST',
      `${FilelistConstants.SESSIONS_URL}/${filelist.id}/directory`,
      undefined,
      onChangeDirectoryChild,
    );

    const renderResult = await renderLayout();
    const {
      getByText,
      findByText,
      filelist1Mocks,
      router,
      userEvent,
      getByPlaceholderText,
      mockStoreListeners,
    } = renderResult;

    const expectFilterInputFocused = () => {
      expect(getByPlaceholderText('Filter...')).toHaveFocus();
    };

    // Check content
    await findByText(clickedDirectory.name);
    expectFilterInputFocused();

    // Go to a child directory
    await userEvent.click(getByText(clickedDirectory.name));
    await waitExpectRequestToMatchSnapshot(onChangeDirectoryChild);

    // Set as loading
    mockStoreListeners.filelist.updated.fire(
      {
        location: {
          ...FilelistGetFilelistItemDirectoryResponse,
          complete: false,
        },
        read: true,
      },
      filelist.id,
    );

    await waitFor(() => {
      expect(router.state.location.state.directory).toEqual(clickedDirectory.path);
    });

    mockStoreListeners.filelist.updated.fire(
      {
        state: FilelistStatePending,
      },
      filelist.id,
    );

    await findByText(FilelistStatePending.str);
    await waitFor(() =>
      expect(filelist1Mocks.table.mockTableManager.isActive()).toBe(false),
    );

    // Set as loaded
    filelist1Mocks.table.mockTableManager.setItems(
      FilelistDirectoryItemsList.items,
      false,
    );
    mockStoreListeners.filelist.updated.fire(
      {
        location: FilelistGetFilelistItemDirectoryResponse,
        read: true,
      },
      filelist.id,
    );

    mockStoreListeners.filelist.updated.fire(
      {
        state: FilelistStateLoaded,
      },
      filelist.id,
    );

    await findByText(FilelistDirectoryItemsList.items[0].name);
    expectFilterInputFocused();

    // Go back to the root
    const onChangeDirectoryRoot = vi.fn();
    server.addRequestHandler(
      'POST',
      `${FilelistConstants.SESSIONS_URL}/${FilelistLoadedResponse.id}/directory`,
      undefined,
      onChangeDirectoryRoot,
    );

    await userEvent.click(getByText('Root'));
    await waitExpectRequestToMatchSnapshot(onChangeDirectoryRoot);

    // Fire the location change event
    mockStoreListeners.filelist.updated.fire(
      {
        location: FilelistGetFilelistItemDirectoryResponse,
        read: true,
      },
      filelist.id,
    );

    filelist1Mocks.table.mockTableManager.setItems(FilelistRootItemsList.items);

    await findByText(FilelistRootItemsList.items[0].name);
    expectFilterInputFocused();

    await navigateToUrl('/', router);
    await filelist1Mocks.table.waitStopped();
  });

  test('should open item details dialog', async () => {
    server.addRequestHandler(
      'GET',
      // eslint-disable-next-line max-len
      `${FilelistConstants.MODULE_URL}/${FilelistLoadedResponse.id}/items/${FilelistGetFilelistItemDirectoryResponse.id}`,
      FilelistGetFilelistItemDirectoryResponse,
    );

    const renderResult = await renderLayout();
    const { findByText, router, findByRole, userEvent, filelist1Mocks } = renderResult;

    const directoryItem = FilelistGetFilelistItemDirectoryResponse;
    await findByText(directoryItem.name);

    // Open details dialog
    const contentCaption = `${directoryItem.name} actions`;
    await openIconMenu(contentCaption, renderResult);
    await clickMenuItem('Details', renderResult);

    const dialog = await findByRole('dialog');

    // Close it
    userEvent.click(within(dialog).getByRole('button', { name: 'Close' }));
    await waitForUrl(`/filelists/session/${FilelistLoadedResponse.id}`, router);

    await navigateToUrl('/', router);
    await filelist1Mocks.table.waitStopped();
  });

  describe('downloads', () => {
    test('should download files', async () => {
      // Download handler
      const onDownloadFile = vi.fn();
      server.addRequestHandler(
        'POST',
        `${QueueConstants.BUNDLES_URL}/file`,
        QueueBundleCreateFileResponse,
        onDownloadFile,
      );

      const renderResult = await renderLayout();
      const { findByText, filelist1Mocks, router } = renderResult;

      filelist1Mocks.table.mockTableManager.setItems(FilelistDirectoryItemsList.items);

      const fileItem = FilelistGetFilelistItemFileResponse;
      await findByText(fileItem.name);

      // Download a file
      const contentCaption = `${fileItem.name} actions`;
      await openIconMenu(contentCaption, renderResult);
      await clickMenuItem('Download', renderResult);

      await waitExpectRequestToMatchSnapshot(onDownloadFile);

      await navigateToUrl('/', router);
      await filelist1Mocks.table.waitStopped();
    });

    test('should download directories', async () => {
      // Download handler
      const onDownloadDirectory = vi.fn();
      server.addRequestHandler(
        'POST',
        FilelistConstants.DIRECTORY_DOWNLOADS_URL,
        FilelistDirectoryDownloadResponse,
        onDownloadDirectory,
      );

      const renderResult = await renderLayout();
      const { findByText, router, filelist1Mocks } = renderResult;

      const directoryItem = FilelistGetFilelistItemDirectoryResponse;
      await findByText(directoryItem.name);

      // Download a directory
      const contentCaption = `${directoryItem.name} actions`;
      await openIconMenu(contentCaption, renderResult);
      await clickMenuItem('Download', renderResult);

      await waitExpectRequestToMatchSnapshot(onDownloadDirectory);
      await navigateToUrl('/', router);
      await filelist1Mocks.table.waitStopped();
    });
  });

  describe('new session', () => {
    const installNewLayoutMocks = (
      newFilelistResponse: API.FilelistSession,
      { mockStoreListeners }: CommonDataMocks,
      createUrl: string = FilelistConstants.SESSIONS_URL,
    ) => {
      server.addRequestHandler(
        'GET',
        getHistoryUrl(HistoryEntryEnum.FILELIST),
        HistoryFilelistResponse,
      );

      installShareProfileMocks(server);
      installUserSearchFieldMocks(server);

      const onSessionCreated = vi.fn(() => {
        mockStoreListeners.filelist.created.fire(newFilelistResponse);
      });
      server.addRequestHandler('POST', createUrl, newFilelistResponse, onSessionCreated);

      return onSessionCreated;
    };

    test('should open local share profile', async () => {
      const renderResult = await renderLayout();
      const {
        queryByText,
        sessionStore,
        router,
        findByText,
        formatter,
        filelistMeMocks,
      } = renderResult;

      const onSessionCreated = installNewLayoutMocks(
        FilelistMeResponse,
        renderResult,
        `${FilelistConstants.SESSIONS_URL}/self`,
      );

      await waitSessionsLoaded(queryByText);

      // Remove existing sessions
      sessionStore.getState().filelists.removeSession(FilelistLoadedResponse);
      sessionStore.getState().filelists.removeSession(FilelistPendingResponse);
      sessionStore.getState().filelists.removeSession(FilelistMeResponse);

      // We should be redirected to the new session layout
      await waitForUrl('/filelists/new', router);

      await findByText('Browse own share...');

      await openMenu('Browse own share...', renderResult);
      await clickMenuItem(
        formatProfileNameWithSize(ShareProfile2, formatter),
        renderResult,
      );

      // We should be redirected to the new session
      await waitFor(() =>
        expect(sessionStore.getState().filelists.activeSessionId).toEqual(
          FilelistMeResponse.id,
        ),
      );

      await findByText(FilelistMeResponse.share_profile.str);

      await waitExpectRequestToMatchSnapshot(onSessionCreated);

      await navigateToUrl('/', router);
      await filelistMeMocks.table.waitStopped();
    });

    test('open remote session', async () => {
      const renderResult = await renderLayout();
      const {
        queryByText,
        sessionStore,
        router,
        findByRole,
        findByText,
        getByText,
        userEvent,
        filelist1Mocks,
      } = renderResult;

      const onSessionCreated = installNewLayoutMocks(
        FilelistLoadedResponse,
        renderResult,
      );

      await waitSessionsLoaded(queryByText);

      // Remove existing sessions
      sessionStore.getState().filelists.removeSession(FilelistLoadedResponse);
      sessionStore.getState().filelists.removeSession(FilelistPendingResponse);
      sessionStore.getState().filelists.removeSession(FilelistMeResponse);

      // We should be redirected to the new session layout
      await waitForUrl('/filelists/new', router);

      // Type the user nick
      const input = await findByRole('combobox');
      await userEvent.type(input, FilelistLoadedResponse.user.nicks);

      // Click the user
      const user1Item = await findByText(FilelistLoadedResponse.user.nicks);
      await userEvent.click(user1Item!);

      // We should be redirected to the new session
      await waitFor(() =>
        expect(sessionStore.getState().filelists.activeSessionId).toEqual(
          FilelistLoadedResponse.id,
        ),
      );

      await waitFor(() =>
        expect(getByText(FilelistLoadedResponse.user.nicks)).toBeTruthy(),
      );

      await waitExpectRequestToMatchSnapshot(onSessionCreated);

      await navigateToUrl('/', router);
      await filelist1Mocks.table.waitStopped();
    });
  });
});
