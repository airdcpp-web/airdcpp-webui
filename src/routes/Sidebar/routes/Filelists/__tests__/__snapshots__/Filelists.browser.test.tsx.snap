// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Filelists > downloads > should download directories 1`] = `
{
  "data": {
    "list_path": "/Ubuntu 20.04 Server/",
    "target_name": "Ubuntu 20.04 Server",
    "user": {
      "cid": "HNLUODI2YZ2U7FDWMHFNJU65ERGKUN4MH7GW5LY",
      "flags": [
        "asch",
        "ccpm",
      ],
      "hub_names": "ADC hub 1 (ADC hub 3)",
      "hub_url": "adcs://hub1.adc/?kp=SHA256/H4DUKGI3U3LQHMZXU6SLQA7LF5RVD6HJFGYZBN2R5RRQ5A2RZ4ZA",
      "hub_urls": [
        "adcs://hub1.adc/?kp=SHA256/H4DUKGI3U3LQHMZXU6SLQA7LF5RVD6HJFGYZBN2R5RRQ5A2RZ4ZA",
        "adcs://hub3.adc:2482",
      ],
      "nicks": "[Hub1]User1 ([Hub3]User1)",
    },
  },
  "method": "POST",
  "path": "filelists/directory_downloads",
}
`;

exports[`Filelists > downloads > should download files 1`] = `
{
  "data": {
    "size": 2877227008,
    "target_name": "ubuntu-20.04.2.0-desktop-amd64.iso",
    "time": 1618062636,
    "tth": "2E2P2K5JP6KARHNFGUMIFENN4SGBENKLPGQNUAA",
    "user": {
      "cid": "HNLUODI2YZ2U7FDWMHFNJU65ERGKUN4MH7GW5LY",
      "flags": [
        "asch",
        "ccpm",
      ],
      "hub_names": "ADC hub 1 (ADC hub 3)",
      "hub_url": "adcs://hub1.adc/?kp=SHA256/H4DUKGI3U3LQHMZXU6SLQA7LF5RVD6HJFGYZBN2R5RRQ5A2RZ4ZA",
      "hub_urls": [
        "adcs://hub1.adc/?kp=SHA256/H4DUKGI3U3LQHMZXU6SLQA7LF5RVD6HJFGYZBN2R5RRQ5A2RZ4ZA",
        "adcs://hub3.adc:2482",
      ],
      "nicks": "[Hub1]User1 ([Hub3]User1)",
    },
  },
  "method": "POST",
  "path": "queue/bundles/file",
}
`;

exports[`Filelists > new session > open remote session 1`] = `
{
  "data": {
    "directory": "/",
    "user": {
      "cid": "HNLUODI2YZ2U7FDWMHFNJU65ERGKUN4MH7GW5LY",
      "hub_url": "adcs://hub1.adc/?kp=SHA256/H4DUKGI3U3LQHMZXU6SLQA7LF5RVD6HJFGYZBN2R5RRQ5A2RZ4ZA",
    },
  },
  "method": "POST",
  "path": "filelists",
}
`;

exports[`Filelists > new session > should open local share profile 1`] = `
{
  "data": {
    "share_profile": 535,
  },
  "method": "POST",
  "path": "filelists/self",
}
`;

exports[`Filelists > should navigate between directories 1`] = `
{
  "data": {
    "list_path": "/Ubuntu 20.04 Server/",
  },
  "method": "POST",
  "path": "filelists/NJSHVYR4ZHCZFUEZNGA7M2D72S5AB4WQAMPBKFA/directory",
}
`;

exports[`Filelists > should navigate between directories 2`] = `
{
  "data": {
    "list_path": "/",
  },
  "method": "POST",
  "path": "filelists/NJSHVYR4ZHCZFUEZNGA7M2D72S5AB4WQAMPBKFA/directory",
}
`;
